include/merlin/memory_pool.hpp

```diff
@@ -0,0 +1,600 @@
+  using async_unique_ptr = std::unique_ptr<type, std::function<void(type*)>>;
+  using shared_ptr = std::shared_ptr<type>;
+
+  inline static sync_unique_ptr make_unique(size_t n) {
+    return sync_unique_ptr(Allocator::alloc(n));
+  }
+
+  inline static async_unique_ptr make_unique(size_t n, cudaStream_t ctx) {
+    return {Allocator::alloc(n, ctx), [ctx](type* p) { Allocator::free(p); }};
+  }
+
+  inline static shared_ptr make_shared(size_t n, cudaStream_t ctx = nullptr) {
+    return {Allocator::alloc(n, ctx),
+            [ctx](type* p) { Allocator::free(p, ctx); }};
+  }
+
+  inline void operator()(type* ptr) { Allocator::free(ptr, nullptr); }
+};
+
+/**
+ * Trivial fallback implementation using the standard C++ allocator. This mostly
+ * exists to ensure interface correctness, and as an illustration of what a
+ * proper allocator implementation should look like.
+ */
+template <class T>
+struct StandardAllocator final : AllocatorBase<T, StandardAllocator<T>> {
+  using type = typename AllocatorBase<T, StandardAllocator<T>>::type;
+
+  static constexpr const char* name{"StandardAllocator"};
+
+  inline static type* alloc(size_t n, cudaStream_t ctx = nullptr) {
+    return new type[n];
+  }
+
+  inline static void free(type* ptr, cudaStream_t ctx = nullptr) {
+    delete[] ptr;
+  }
+};
+
+/**
+ * Claim/release buffers in pinned host memory.
+ */
+template <class T>
+struct HostAllocator final : AllocatorBase<T, HostAllocator<T>> {
+  using type = typename AllocatorBase<T, HostAllocator<T>>::type;
+
+  static constexpr const char* name{"HostAllocator"};
+
+  inline static type* alloc(size_t n, cudaStream_t ctx = nullptr) {
+    void* ptr;
+    CUDA_CHECK(cudaMallocHost(&ptr, sizeof(T) * n));
+    return reinterpret_cast<type*>(ptr);
+  }
+
+  inline static void free(type* ptr, cudaStream_t ctx = nullptr) {
+    CUDA_CHECK(cudaFreeHost(ptr));
+  }
+};
+
+/**
+ * Claim/release buffers in the active CUDA device. Will not test if the correct
+ * device was used, and throw if CUDA runtime API response is negative.
+ */
+template <class T>
+struct DeviceAllocator final : AllocatorBase<T, DeviceAllocator<T>> {
+  using type = typename AllocatorBase<T, DeviceAllocator<T>>::type;
+
+  static constexpr const char* name{"DeviceAllocator"};
+
+  inline static type* alloc(size_t n, cudaStream_t ctx = nullptr) {
+    void* ptr;
+    cudaError_t res;
+    if (ctx) {
+      res = cudaMallocAsync(&ptr, sizeof(T) * n, ctx);
+    } else {
+      res = cudaMalloc(&ptr, sizeof(T) * n);
+    }
+    CUDA_CHECK(res);
+    return reinterpret_cast<type*>(ptr);
+  }
+
+  inline static void free(type* ptr, cudaStream_t ctx = nullptr) {
+    cudaError_t res;
+    if (ctx) {
+      res = cudaFreeAsync(ptr, ctx);
+    } else {
+      res = cudaFree(ptr);
+    }
+    CUDA_CHECK(res);
+  }
+};
+
+/**
+ * Wrapper around another allocator that prints debug messages.
+ */
+template <class Allocator>
+struct DebugAllocator final
```

Comment:
Move `DebugAllocator` into the `memory_pool_test.cpp`

Reply:
Not a fan. But done!